smallzstd/zstd.go

```diff
@@ -0,0 +1,68 @@
+// package smallzstd produces zstd encoders and decoders optimized for
+// low memory usage, at the expense of compression efficiency.
+package smallzstd
+
+import (
+	"io"
+
+	"github.com/klauspost/compress/zstd"
+)
+
+// This package is optimized primarily for the memory cost of
+// compressing and decompressing data. We reduce this cost in two
+// major ways: disable parallelism within the library (i.e. don't use
+// multiple CPU cores to decompress), and drop the compression window
+// down from the defaults of 4-16MiB, to 8kiB.
+//
+// Decompressors cost 2x the window size in RAM to run, so by using an
+// 8kiB window, we can run ~1000 more decompressors per unit of memory
+// than with the defaults.
+//
+// Depending on context, the benefit is either being able to run more
+// decoders (e.g. in our logs processing system), or having a lower
+// memory footprint when using compression in network protocols
+// (e.g. in tailscaled, which should have a minimal RAM cost).
+
+const WindowSize = 8 << 10 // 8kiB
+
+func NewSmallDecoder(r io.Reader, options ...zstd.DOption) (*zstd.Decoder, error) {
```

Comment:
the name "Small" in these funcs are stuttery with the package name (classic https://golang.org/doc/effective_go.html#package-names example)